rust/src/dataset/cleanup.rs

```diff
@@ -0,0 +1,697 @@
+                let relative_data_path = remove_prefix(&full_data_path, &self.dataset.base);
+                invalid_files.unref_data_paths.remove(&relative_data_path);
+            }
+            let delpath = fragment
+                .deletion_file
+                .as_ref()
+                .map(|delfile| deletion_file_path(&self.dataset.base, fragment.id, delfile));
+            if delpath.is_some() {
+                let relative_path = remove_prefix(&delpath.unwrap(), &self.dataset.base);
+                invalid_files.unref_delete_paths.remove(&relative_path);
+            }
+        }
+        let full_path = join_paths(&self.dataset.base, manifest_path);
+        let indexes =
+            read_manifest_indexes(&self.dataset.object_store, &full_path, manifest).await?;
+        for index in indexes {
+            let uuid_str = index.uuid.to_string();
+            invalid_files.unref_index_uuids.remove(&uuid_str);
+        }
+        Ok(())
+    }
+
+    fn discover_path(&mut self, path: &Path, files_list: &mut FilesList) -> Result<()> {
+        let relative_path = remove_prefix(path, &self.dataset.base);
+        if relative_path.as_ref().starts_with("_indices") {
+            // For indices we just grab the UUID because we want to delete the entire
+            // folder if we determine an index is expired.
+            if let Some(uuid) = relative_path.parts().nth(1) {
+                files_list
+                    .index_uuids
+                    .insert(Path::parse(uuid)?.as_ref().to_owned());
+            }
+            // Intentionally returning early here.  We don't want data files in index directories
+            // to be treated like normal data files.
+            return Ok(());
+        }
+        match path.extension() {
+            Some("lance") => {
+                if relative_path.as_ref().starts_with("data") {
+                    files_list.data_file_paths.push(relative_path);
+                }
+            }
+            Some("manifest") => {
+                // We intentionally ignore _latest.manifest.  We should never delete
+                // it and we can assume that it is a clone of one of the _versions/... files
+                if relative_path.as_ref().starts_with("_versions") {
+                    files_list.manifest_file_paths.push(relative_path)
+                }
+            }
+            Some("arrow") => {
+                if relative_path.as_ref().starts_with("_deletions") {
+                    files_list.delete_file_paths.push(relative_path);
+                }
+            }
+            Some("bin") => {
+                if relative_path.as_ref().starts_with("_deletions") {
+                    files_list.delete_file_paths.push(relative_path);
+                }
+            }
+            _ => (),
+        };
+        Ok(())
+    }
+}
+
+pub async fn cleanup_old_versions(dataset: &Dataset, before: DateTime<Utc>) -> Result<()> {
```

Comment:
Could we add some docs? We should set expectations about which files are deleted and which are ignored.

Also, perhaps we should emit some statistics, like the number of files that were deleted and number of versions deleted?

Reply:
I've added a comment.  The function now returns some stats about what it deleted.